test/unit/org/apache/cassandra/io/filesystem/ListenableFileSystem.java

```diff
@@ -0,0 +1,510 @@
+        {
+            long position = position();
+            notifyListeners(onWrite, l -> l.preWrite(path, this, position, src));
+            int write = super.write(src);
+            notifyListeners(onWrite, l -> l.postWrite(path, this, position, src, write));
+            return write;
+        }
+
+        @Override
+        public int write(ByteBuffer src, long position) throws IOException
+        {
+            notifyListeners(onWrite, l -> l.preWrite(path, this, position, src));
+            int write = super.write(src, position);
+            notifyListeners(onWrite, l -> l.postWrite(path, this, position, src, write));
+            return write;
+        }
+
+        @Override
+        public FileChannel position(long newPosition) throws IOException
+        {
+            long position = position();
+            notifyListeners(onChannelMeta, l -> l.prePosition(path, this, position, newPosition));
+            super.position(newPosition);
+            notifyListeners(onChannelMeta, l -> l.postPosition(path, this, position, newPosition));
+            return this;
+        }
+
+        @Override
+        public FileChannel truncate(long size) throws IOException
+        {
+            long currentSize = this.size();
+            notifyListeners(onChannelMeta, l -> l.preTruncate(path, this, currentSize, size));
+            super.truncate(size);
+            long latestSize = this.size();
+            notifyListeners(onChannelMeta, l -> l.postTruncate(path, this, currentSize, size, latestSize));
+            return this;
+        }
+
+        @Override
+        public void force(boolean metaData) throws IOException
+        {
+            notifyListeners(onChannelMeta, l -> l.preForce(path, this, metaData));
+            super.force(metaData);
+            notifyListeners(onChannelMeta, l -> l.postForce(path, this, metaData));
+        }
+
+        @Override
+        public long transferTo(long position, long count, WritableByteChannel target) throws IOException
+        {
+            notifyListeners(onTransferTo, l -> l.preTransferTo(path, this, position, count, target));
+            long transfered = super.transferTo(position, count, target);
+            notifyListeners(onTransferTo, l -> l.postTransferTo(path, this, position, count, target, transfered));
+            return transfered;
+        }
+
+        @Override
+        public long transferFrom(ReadableByteChannel src, long position, long count) throws IOException
+        {
+            notifyListeners(onTransferFrom, l -> l.preTransferFrom(path, this, src, position, count));
+            long transfered = super.transferFrom(src, position, count);
+            notifyListeners(onTransferFrom, l -> l.postTransferFrom(path, this, src, position, count, transfered));
+            return transfered;
+        }
+
+        @Override
+        public MappedByteBuffer map(MapMode mode, long position, long size) throws IOException
```

Comment:
since this is based off the buffer cache... we "might" be able to get this to be "correct" by having the read/write methods change behavior when a buffer exists for the provided offset?  I didn't go down this path as nothing in C* looked to need that complexity

Reply:
YAGNI